lua/plenary/collections/py_list.lua

```diff
@@ -0,0 +1,157 @@
+-- vim:sw=2
+local List = {}
+
+setmetatable(List, List)
+
+function List:__index(key)
+  if self ~= List then
+    local field = List[key]
+    if field then return field end
+  end
+end
+
+function List:__call(tbl)
+  if type(tbl) == 'table' then
+    local len = #tbl
+    local obj = setmetatable(tbl, self)
+    obj._len = len
+    return obj
+  end
+  error 'List constructor must be called with table argument'
+end
+
+function List:__tostring()
+  if rawequal(self, List) then return '<List class>' end
+  if #self == 0 then return '[]' end
+  local result = {'['}
+  for _, v in ipairs(self) do
+    local repr = tostring(v)
+    if type(v) == 'string' then repr = '"' .. repr .. '"' end
+    result[#result + 1] = repr
+    result[#result + 1] = ', '
+  end
+  result[#result] = ']'
+  return table.concat(result)
+end
+
+function List:__eq(other)
+  if #self ~= #other then return false end
+  for i = 1, #self do if self[i] ~= other[i] then return false end end
+  return true
+end
+
+function List:__mul(other)
+  local result = List {}
+  for i = 1, other do result[i] = self end
+  return result
+end
+
+function List:__len()
+  return self._len
+end
+
+function List:__concat(other)
+  local result = List {}
+  for _, v in ipairs(self) do result[#result + 1] = v end
+  for _, v in ipairs(other) do result[#result + 1] = v end
+  return result
+end
+
+function List.is_list(tbl)
+  local meta = getmetatable(tbl) or {}
+  return meta == List
+end
+
+function List:append(other)
```

Comment:
I would call this `push` because you called the other one `pop`. Then add an `insert` method that will take an index.